### ISSUE TITLE
how to map certain fields to custom / open ended field types?

### ISSUE THREAD

**Issue by RutledgePaulV** [RutledgePaulV/rest-query-engine#3]
transferring from: https://github.com/RutledgePaulV/q-builders/issues/11

dear Paul,

today we were running some queries using the following RSQL:

```sql
chemicalCompound.metaData=q='name=="total exact mass" and value=gt=411.31 and value=lt=411.4'
```

which generated the following mongo query

```json
{ "chemicalCompound.metaData" : { "$elemMatch" : { "$and" : [ { "name" : "total exact mass"} , { "value" : { "$gt" : "411.31"}} , { "value" : { "$lt" : "411.4"}}]}}}
```

which is great!

Sadly it mapped the numbered arguments as String instead of Double, which caused the backend not to find any data. The correct query, would have been:

```json
{ "chemicalCompound.metaData" : { "$elemMatch" : { "$and" : [ { "name" : "total exact mass"} , { "value" : { "$gt" : 411.31}} , { "value" : { "$lt" : 411.4}}]}}}
```

Is there some way, to specify in the pipeline. To map our results?

I have the feeling that this is related to our Metadata Object not specifying a type (Sorry it's scala)

```scala
case class MetaData(
category: String ,
computed: Boolean,
deleted: Boolean,
hidden: Boolean,
name: String,
score: Score,
unit: String,
url: String,
@JsonDeserialize(using = classOf[NumberDeserializer])
value: Any
)
```

It can be any object, since we do not force the user to specify if it's a double or a string,etc. This is related to our data matrix

And we have a custom mapper, to serialize this based on the following logic

```scala
class NumberDeserializer extends JsonDeserializer[Any] {

override def deserialize(jsonParser: JsonParser, deserializationContext: DeserializationContext): Any = {

try {
  val jsonNode: JsonNode = jsonParser.getCodec.readTree(jsonParser)

  val content = jsonNode.textValue

  if (content != null) {
    if (content.toLowerCase.equals("true")) {
      true
    }
    else if (content.toLowerCase().equals("false")) {
      false
    }
    else {
      try {
        content.toInt
      } catch {
        case e: NumberFormatException => try {
          content.toDouble
        }
        catch {
          case e2: NumberFormatException => content
        }
      }
    }
  }
  else{
    content
  }
}
catch {
  case e: Exception => e.printStackTrace()
}
}
}
```

Which basically tries to figure out the correct value from the given JSON for our object. So that our queries can be executed by number.

Is it possible that we can register such a mapper somewhere in your code, to be able to map our unusual configuration?

thanks again!

**Comment by RutledgePaulV** [comment]
@berlinguyinca, excellent question. I had planned support for stuff like this but hadn't gotten around to it until 2 minutes ago :)

Please see my brand new test based on your use case. It's a full example of how to use it:

https://github.com/RutledgePaulV/rest-query-engine/blob/develop/src/test/java/com/github/rutledgepaulv/rqe/pipes/DemoCustomConversionPipeline.java


Also, I had to make a couple adjustments to make this possible, so I'll release a new snapshot version for you here in a couple minutes. It will be 0.6-SNAPSHOT

**Comment by RutledgePaulV** [comment]
One thing to watch out for is the "supports" method of your custom converter. Keep in mind that it might be coming in via a few different routes.

Whereas an "elemMatch" will strip of the leading "chemicalCompound.metaData" from "value", if you instead queried directly against the field with "chemicalCompound.metaData.value" you'd need your converter to return true for a different property path / class combo.

**Comment by RutledgePaulV** [comment]
I actually decided to fix this another way as well. I added a "best effort" converter for whenever the pipeline finds itself trying to convert a string value (as parsed by RSQL) into a field whose type just resolves to "Object / Any". This best effort converter tries booleans, longs, doubles, dates in iso format, and defaults to just the string as is if none of the others succeeded. This is probably the easiest to use (should only require updating to 0.6.1-SNAPSHOT), but you're still welcome to use the custom converter functionality if you need it / want it!

**Comment by berlinguyinca** [comment]
Thank you and thats great and seems to generate the queries as required.

On Mon, Mar 7, 2016 at 9:30 PM, Paul Rutledge <notifications@github.com>
wrote:

> I actually decided to fix this another way as well. I added a "best
> effort" converter for whenever the pipeline finds itself trying to convert
> a string value (as parsed by RSQL) into a field whose type just resolves to
> "Object / Any". This best effort converter tries booleans, longs, doubles,
> dates in iso format, and defaults to just the string as is if none of the
> others succeeded. This is probably the easiest to use (should only require
> updating to 0.6.1-SNAPSHOT), but you're still welcome to use the custom
> converter functionality if you need it / want it!
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/RutledgePaulV/rest-query-engine/issues/3#issuecomment-193616199>
> .
>



-- 
------------------------------------------------------------
Lead Developer - Fiehnlab, UC Davis

gert wohlgemuth

work:
http://fiehnlab.ucdavis.edu/staff/wohlgemuth

phone:
530 665 9477

coding blog:
http://codingandmore.blogspot.com

linkedin:
http://www.linkedin.com/profile/view?id=28611299&trk=tab_pro

**Comment by RutledgePaulV** [comment]
Awesome! Feel free to create more tickets if anything comes up or you find yourself wishing it did something that it doesn't.

**Comment by berlinguyinca** [comment]
thanks and will do!

g.

On Tue, Mar 8, 2016 at 1:43 PM, Paul Rutledge <notifications@github.com>
wrote:

> Closed #3 <https://github.com/RutledgePaulV/rest-query-engine/issues/3>.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/RutledgePaulV/rest-query-engine/issues/3#event-582395482>
> .
>



-- 
------------------------------------------------------------
Lead Developer - Fiehnlab, UC Davis

gert wohlgemuth

work:
http://fiehnlab.ucdavis.edu/staff/wohlgemuth

phone:
530 665 9477

coding blog:
http://codingandmore.blogspot.com

linkedin:
http://www.linkedin.com/profile/view?id=28611299&trk=tab_pro
